dali/imgcodec/decoders/tiff_libtiff.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tiffio.h>
+#include "dali/imgcodec/decoders/tiff_libtiff.h"
+#include "dali/imgcodec/util/convert.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/kernels/common/utils.h"
+
+#define LIBTIFF_CALL_SUCCESS 1
+#define LIBTIFF_CALL(call)                                \
+  do {                                                    \
+    int retcode = (call);                                 \
+    DALI_ENFORCE(LIBTIFF_CALL_SUCCESS == retcode,         \
+      "libtiff call failed with code "                    \
+      + std::to_string(retcode) + ": " #call);            \
+  } while (0)
+
+namespace dali {
+namespace imgcodec {
+
+namespace detail {
+
+class DecoderHelper {
+ private:
+  std::shared_ptr<InputStream> stream;
+  const void *buffer;
+  size_t buffer_size;
+
+ public:
+  explicit DecoderHelper(ImageSource *in) : stream(in->Open()) {
+    if (in->Kind() == InputKind::HostMemory) {
+      buffer = in->RawData();
+      buffer_size = in->Size();
+    } else {
+      buffer = nullptr;
+    }
+  }
+
+  static tmsize_t read(thandle_t handle, void *buffer, tmsize_t n) {
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    return helper->stream->Read(buffer, n);
+  }
+
+  static tmsize_t write(thandle_t, void *, tmsize_t) {
+    // Not used for decoding.
+    return 0;
+  }
+
+  static toff_t seek(thandle_t handle, toff_t offset, int whence) {
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    helper->stream->SeekRead(offset, whence);
+    return helper->stream->TellRead();
+  }
+
+  static int map(thandle_t handle, void **base, toff_t *size) {
+    DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
+    *base = const_cast<void*>(helper->buffer);
```

Comment:
What happens if `in->Kind() != InputKind::HostMemory`? 

I am not sure where do we use it, but in that case we return unitialized bufer_size and valid buffer ptr.
Btw, why can't we extract those two things from stream here?

Reply:
Oh, I see below, that we are not providing the DecodeHelper::map in any other case. 

Can you maybe add some doc about it? It would be nice to have a check against using other kinds here, but probably that's overkill. Maybe assert?

Reply:
> Btw, why can't we extract those two things from stream here?
Good idea. Now I keep the `ImageSource` in a field and get the buffer pointer/size inside `map`

Reply:
I added a comment and a check for using only `HostMem` here. This function will be called once per image, so it shouldn't hurt